internal/core/data/init.go

```diff
@@ -51,32 +50,36 @@ func NewBootstrap(router *mux.Router) *Bootstrap {
 }
 
 // BootstrapHandler fulfills the BootstrapHandler contract and performs initialization needed by the data service.
-func (b *Bootstrap) BootstrapHandler(_ context.Context, _ *sync.WaitGroup, _ startup.Timer, dic *di.Container) bool {
+func (b *Bootstrap) BootstrapHandler(ctx context.Context, wg *sync.WaitGroup, _ startup.Timer, dic *di.Container) bool {
 	loadRestRoutes(b.router, dic)
 
-	lc := container.LoggingClientFrom(dic.Get)
 	configuration := dataContainer.ConfigurationFrom(dic.Get)
-
-	// initialize clients required by service.
 	registryClient := container.RegistryFrom(dic.Get)
-	mdc := metadata.NewDeviceClient(
-		types.EndpointParams{
-			ServiceKey:  clients.CoreMetaDataServiceKey,
-			Path:        clients.ApiDeviceRoute,
-			UseRegistry: registryClient != nil,
-			Url:         configuration.Clients["Metadata"].Url() + clients.ApiDeviceRoute,
-			Interval:    configuration.Service.ClientMonitor,
-		},
-		endpoint.Endpoint{RegistryClient: &registryClient})
-	msc := metadata.NewDeviceServiceClient(
-		types.EndpointParams{
-			ServiceKey:  clients.CoreMetaDataServiceKey,
-			Path:        clients.ApiDeviceServiceRoute,
-			UseRegistry: registryClient != nil,
-			Url:         configuration.Clients["Metadata"].Url() + clients.ApiDeviceRoute,
-			Interval:    configuration.Service.ClientMonitor,
-		},
-		endpoint.Endpoint{RegistryClient: &registryClient})
+
+	mdc := metadata.NewDeviceClient(urlclient.New(
+		registryClient != nil,
+		endpoint.New(
+			ctx,
+			wg,
+			&registryClient,
```

Comment:
What if registryClient == nil?   &nil?

Occurs elsewhere.

Reply:
Yes. We build a local client and the nil pointer is never examined, as intended.

Reply:
I'm not communicating effectively.  If `registryClient` has a nil value, what does `&registryClient` get you?

Reply:
[A pointer to nil, as intended.](https://play.golang.org/p/kZgi90aRklh)

Reply:
![image](https://user-images.githubusercontent.com/464327/74482730-b18ce900-4e7a-11ea-941c-8ff2a3b68482.png)

By your example, it's a pointer to a pointer to nil.  Yes it's technically correct code.  It's also awkward code.